text/0000-simd-infrastructure.md

```diff
@@ -0,0 +1,422 @@
+enforce a specific nominal type.
+
+NB. The structural typing is just for the declaration: if a SIMD intrinsic
+is declared to take a type `X`, it must always be called with `X`,
+even if other types are structurally equal to `X`. Also, within a
+signature, SIMD types that must be structurally equal must be nominally
+equal. I.e. if the `add_...` all refer to the same intrinsic to add a
+SIMD vector of bytes,
+
+```rust
+// (same length)
+struct A(u8, u8, ..., u8);
+struct B(u8, u8, ..., u8);
+
+extern "platform-intrinsic" {
+    fn add_aaa(x: A, y: A) -> A; // ok
+    fn add_bbb(x: B, y: B) -> B; // ok
+    fn add_aab(x: A, y: A) -> B; // error, expected B, found A
+    fn add_bab(x: B, y: A) -> B; // error, expected A, found B
+}
+
+fn double_a(x: A) -> A {
+    add_aaa(x, x)
+}
+fn double_b(x: B) -> B {
+    add_aaa(x, x) // error, expected A, found B
+}
+```
+
+There would additionally be a small set of cross-platform operations
+that are either generally efficiently supported everywhere or are
+extremely useful. These won't necessarily map to a single instruction,
+but will be shimmed as efficiently as possible.
+
+- shuffles and extracting/inserting elements
+- comparisons
+- arithmetic
+- conversions
+
+All of these intrinsics are imported via an `extern` directive similar
+to the process for pre-existing intrinsics like `transmute`, however,
+the SIMD operations are provided under a special ABI:
+`platform-intrinsic`. Use of this ABI (and hence the intrinsics) is
+initially feature-gated under the `platform_intrinsics` feature
+name. Why `platform-intrinsic` rather than say `simd-intrinsic`? There
+are non-SIMD platform-specific instructions that may be nice to expose
+(for example, Intel defines an `_addcarry_u32` intrinsic corresponding
+to the `ADC` instruction).
+
+### Shuffles & element operations
+
+One of the most powerful features of SIMD is the ability to rearrange
+data within vectors, giving super-linear speed-ups sometimes. As such,
+shuffles are exposed generally: intrinsics that represent arbitrary
+shuffles.
+
+This may violate the "one instruction per instrinsic" principal
+depending on the shuffle, but rearranging SIMD vectors is extremely
+useful, and providing a direct intrinsic lets the compiler (a) do the
+programmers work in synthesising the optimal (short) sequence of
+instructions to get a given shuffle and (b) track data through
+shuffles without having to understand all the details of every
+platform specific intrinsic for shuffling.
+
+```rust
+extern "platform-intrinsic" {
+    fn simd_shuffle2<T, Elem>(v: T, w: T, i0: u32, i1: u32) -> Simd2<Elem>;
+    fn simd_shuffle4<T, Elem>(v: T, w: T, i0: u32, i1: u32, i2: u32, i3: u32) -> Sidm4<Elem>;
+    fn simd_shuffle8<T, Elem>(v: T, w: T,
+                              i0: u32, i1: u32, i2: u32, i3: u32,
+                              i4: u32, i5: u32, i6: u32, i7: u32) -> Simd8<Elem>;
+    fn simd_shuffle16<T, Elem>(v: T, w: T,
```

Comment:
There's no fundamental reason for it that I know of (although I'm not 100% sure if the compiler currently supports the necessary const-eval stuff, but that should just be an implementation detail). In any case, your approach does seem like it may be better.